File: minimum_falling_path_sum.py
# Time Complexity: O(n) were n is the number of rows in the matrix.
# Space Complexity: O(1)
#  Did this code successfully run on Leetcode : Yes
#  Any problem you faced while coding this : No

class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        result = 0
        for i in range(len(matrix) - 2, -1, -1):
            for j in range(len(matrix[0])):
                if j==0:
                    matrix[i][j] = matrix[i][j]+  min(matrix[i+1][j], matrix[i+1][j+1])
                elif j == len(matrix[0]) -1:
                    matrix[i][j] = matrix[i][j] + min(matrix[i+1][j-1], matrix[i+1][j])
                else:
                    matrix[i][j] = matrix[i][j] +  min(matrix[i+1][j-1], matrix[i+1][j], matrix[i+1][j+1])
        return min(matrix[0])

File: minimum-falling-path-sum.java
/* Time Complexity: O(m*n)
 * Space Complexity:O(m)
 * Did this code successfully run on Leetcode : yes
 * Any problem you faced while coding this : No
 */

class Solution {
    public int minFallingPathSum(int[][] matrix) {
        //if only one subarray
        if(matrix.length == 1) return matrix[0][0];
        int n = matrix.length;
        int[] arr = new int[n];
        //copying values from last row
        for(int j=0;j<n;j++) {
            arr[j] = matrix[n-1][j];
        }
        int curr = 0;
        int prev = 0;
        int min = Integer.MAX_VALUE;
        for(int i=n-2;i>=0;i--) {
            for(int j=0;j<n;j++) {
                //incase of first index, we got two possiblity, below or diagonally right
                if(j == 0) {
                    curr = arr[j];
                    arr[j] = matrix[i][j] + Math.min(arr[j], arr[j+1]);
                    prev = curr;

                } else if(j==n-1) {
                    //incase of last index, we got two possiblity, below or diagonally left
                    arr[j] = matrix[i][j] + Math.min(prev, arr[j]);
                } else {
                    //incase of middle index, we got three possiblities, below or diagonally left/right
                    curr = arr[j];
                    arr[j] = matrix[i][j]+Math.min( prev, Math.min(arr[j],arr[j+1]));
                    prev = curr;
                }
                if(i==0)
                    min = Math.min(min, arr[j]);
            }
        }
        return min;
    }
}